### DISCUSSION TITLE
Trying to understand auto cancellation feature from JS-SDK

### DISCUSSION THREAD

**Discussion by speryt** [pocketbase/pocketbase#4843] [Q&A] [answered]
I'm trying to understand auto cancellation feature of JS-SDK. I mean I understand that this is preventing same requests from happening and just execute the last one. But I have 2 major questions about that:
1. What is the actual purpose of this? To reduce the number of requests to the PocketBase? Some sort of caching mechanism?
2. Is it safe to disable it, or it's better to keep it on and work around it?

In my SvelteKit app with SSR rendering I have following use case:
I'm loading paginated data from PB and provide user with pagination controls with which they can change which page should be loaded. This is happening on the fronted while actual request is happening on the server using JS-SDK. If I understand it correctly (haven't tested it yet) if 2 users will at the same time select the same page to display, from the server point of view they will both request the same resource so I'll have one of those requests canceled - right?

The only thing I've observed so far is that every now and then I'm getting HTTP 500 error when loading data which seems to be related to some race condition when I'm changing pages too fast. Although I'm not 100% sure JS-SDK and PB are responsible for that. It might be my code as well.

Anyhow - is there any recommendation how this auto cancelation should be handled? At the moment I'm returning empty list of elements but I believe it's clear that it isn't ideal.

**Comment by ganigeorgiev** [comment]
> What is the actual purpose of this?

It is intended to prevent unhandled concurrent Promises, for example clicking on a button several times with each click firing a new request - there is no guarantee that the first click will resolve before the second one which could cause unintended side effects in your app if you are updating some local state after a successful request.

> Is it safe to disable it, or it's better to keep it on and work around it?

The autocancellation has no relation to the security - it just helps preventing async bugs.

> if 2 users will at the same time select the same page to display, from the server point of view they will both request the same resource so I'll have one of those requests canceled - right?

I'm not sure I follow. In a JS SSR context is expected each request to have its own unique JS SDK instance, otherwise you are risking requests from different users to write over the same `pb.authStore` (_see also [SSR integration](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration)_). Each JS SDK instance has its own separate autocancellation tracking.

It is safe to have a global JS SDK instance in a server context **only** when using the SDK as admin, aka. you are not operating with the individual users auth state. In that particular case you have to disable the autocancellation because you'll want to process concurrent requests. See an example admin client in https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7621130.


> The only thing I've observed so far is that every now and then I'm getting HTTP 500 error when loading data

PocketBase currently doesn't throw 500 errors so it is most likely something related to your code but without details about the exact error message and code that triggers it I don't know how to help.

> But anyhow - is there any recommendation how this auto cancelation should be handled? 

There is no universal answer to your question as it will depend on what is the desired behavior. In general to handle the autocancellation you have to wrap your request in a `try/catch` and check if `err.isAbort`. Usually if `err.isAbort` is `true` you can silently ignore the call (aka. empty `return`) and leave the other pending request to complete and update your state.

**Reply by speryt** [reply]
> The autocancellation has no relation to the security - it just helps preventing async bugs.

Sorry, I wasn't clear enough. I didn't ask about security but rather safety as in preventing some major issues as an async bugs you've mentioned.
> It is safe to have a global JS SDK instance in a server context **only** when using the SDK as admin, aka. you are not operating with the individual users auth state. In that particular case you have to disable the autocancellation because you'll want to process concurrent requests. See an example admin client in [#3792 (reply in thread)](https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7621130).

I think this is my use case (at the moment at least). I don't require individual users to login and based on that retrieve data from database. All of the records are available for all of the users. I'm actually using right now special "faceless" account to authenticate with SDK. I couldn't find any difference between authenticating as an admin vs as a special user which is added to API rule for the collection.
I'll take a look at the link you pointed.
> There is no universal answer to your question as it will depend on what is the desired behavior. In general to handle the autocancellation you have to wrap your request in a `try/catch` and check if `err.isAbort`. Usually if `err.isAbort` is `true` you can silently ignore the call (aka. empty `return`) and leave the other pending request to complete and update your state.

This is what I started to do (returning empty result on error), but I can sometime end up with empty list as a final request return. And I'm not sure how to cope such cases. I'll need to test it few more times to better catch moment when it is failing.

**Comment by speryt** [comment]
One more question to your comments from #3792 . You are saying there

> PocketBase is designed to be accessed directly. What would be the benefit of proxying every request/message through the node server?

So do I understand correctly that you are assuming that each user of the website will be logged in and will be sing this credentials to access PB directly without going through server? How is it supposed to work with cases where user is not required to login? Let say e-commerce website with catalog of products (before any purchase transaction is happening), or blog.

This is use case I am having as mentioned earlier.

**Reply by ganigeorgiev** [reply]
> How is it supposed to work with cases where user is not required to login? 

If your content is public and doesn't require authentication then you can simply leave the API rule empty.

I'm honestly getting tired by this SSR related questions and I really don't want to deal with this type of support anymore.

If you want to use JS SSR (ex. you need SEO) then there is nothing wrong with this, you are free to do it as long as you understand what is your execution flow and that you have 2 backends to maintain and eventually sync (the node server and PocketBase).

PocketBase was designed for my personal use case (https://github.com/presentator/presentator/issues/183) and interacting with the PocketBase server from a browser is the easiest way to use it.

**Reply by speryt** [reply]
Ok, got it, thanks.
In short then - PocketBase is intended to be used on a client side and not on a server side.
All of the server side frameworks (SvelteKit etc.) are just additions which might work.

Personally, I don't think this is safe to provide to the user URL of our database. Which I believe is the case if we'll just go with pure Svelte. Unless I'm missing some step that will hide it from the user. But well, it's not the question I've been looking answer for. Just a comment.

**Reply by ganigeorgiev** [reply, marked as answer]
> Personally, I don't think this is safe to provide to the user URL of our database

PocketBase is not just a database, it is a backend. That's why the collections have API rules. The goal is to not have to deal with another backend.

In any case I'll stop engaging to avoid further confusion.

**Comment by nomadshiba** [comment]
I never understood the purpose of this, aborting requests is the responsibility of the front-end.
Making it error like this on the api client level will just make you ignore the logic issues in the code.

Also this can randomly break code bases where many components fetch some data async randomly, and there is a chance something will be asked twice by two different components before the first request resolves. In those cases you are suppose to clone the same response for both requests, not cancel the previous one, which just causes unexpected errors.

**Reply by ganigeorgiev** [reply]
> aborting requests is the responsibility of the front-end.

And it is exactly what we do. PocketBase was designed with client-side first usage in mind. If you haven't already please read https://pocketbase.io/docs/how-to-use/.

> Making it error like this on the api client level will just make you ignore the logic issues in the code.

It is usually exactly the opposite - it forces you to fix them.

> Also this can randomly break code bases where many components fetch some data async randomly, and there is a chance something will be asked twice by two different components before the first request resolves

And this usually indicates inefficiencies in the client code base and its API requests usage.

In any case, I don't want to argue and I don't think I can provide anymore arguments than what is already written in my previous replies. If autocancellation is causing you trouble then you can simply disable it with `pb.autoCancellation(false)`. For more details and other options you can also read https://github.com/pocketbase/js-sdk?tab=readme-ov-file#auto-cancellation.
